lib/Utils/Logger.cpp

```diff
@@ -51,74 +47,166 @@ std::string toString(const SeverityLevel &Level) {
+          Filename, EC,
+          llvm::sys::fs::OpenFlags::OF_Append |
+              llvm::sys::fs::OpenFlags::OF_ChildInherit);
+    } else {
+      LogfileStreams[Filename] = std::make_unique<llvm::raw_fd_ostream>(
+          Filename, EC, llvm::sys::fs::OpenFlags::OF_ChildInherit);
+    }
+    // Following
+    // https://stackoverflow.com/questions/41699343/how-do-i-test-that-an-stderror-code-is-not-an-error
+    if (EC) {
+      LogfileStreams[Filename] = nullptr;
+      llvm::errs() << "Failed to open logfile: " << Filename << '\n';
+      llvm::errs() << EC.message() << '\n';
+      return false;
+    }
+  }
+  return true;
 }
 
-void logFormatter(const boost::log::record_view &View,
-                  boost::log::formatting_ostream &OS) {
-#ifdef DYNAMIC_LOG
-  // OS << View.attribute_values()["LineCounter"].extract<int>() << " "
-  OS << View.attribute_values()["Duration"]
-            .extract<boost::posix_time::ptime::time_duration_type>()
-     //<< " "
-     // View.attribute_values()["Timestamp"].extract<boost::posix_time::ptime>()
-     << " - [" << View.attribute_values()["Severity"].extract<SeverityLevel>()
-     << "] " << View.attribute_values()["Message"].extract<std::string>();
-#endif
+llvm::raw_ostream &
+Logger::getLogStream(const std::optional<SeverityLevel> &Level,
+                     const std::optional<llvm::StringRef> &Category) {
+  if (Category.has_value()) {
+    auto CategoryLookupIt = CategoriesToStreamVariant.find(Category.value());
+    if (CategoryLookupIt == CategoriesToStreamVariant.end()) {
+      return llvm::nulls();
+    }
+    return getLogStream(Level, CategoryLookupIt->second);
+  }
+  return getLogStream(Level, LevelsToStreamVariant);
 }
 
-void initializeLogger(bool UseLogger, const string &LogFile) {
-#ifdef DYNAMIC_LOG
-  // Using this call, logging can be enabled or disabled
-  boost::log::core::get()->set_logging_enabled(UseLogger);
-  using text_sink = boost::log::sinks::synchronous_sink<
-      boost::log::sinks::text_ostream_backend>;
-  boost::shared_ptr<text_sink> Sink = boost::make_shared<text_sink>();
-  boost::shared_ptr<std::ostream> Stream =
-      [](const string &LogFile) -> boost::shared_ptr<std::ostream> {
-    if (LogFile.empty()) {
-      // the easiest way is to write the logs to std::clog
-      return boost::shared_ptr<std::ostream>(&std::clog, boost::null_deleter{});
+llvm::raw_ostream &
+Logger::getLogStream(const std::optional<SeverityLevel> &Level,
+                     const std::map<std::optional<SeverityLevel>,
+                                    std::variant<StdStream, std::string>>
+                         &PassedLevelsToStreamVariant) {
+  if (Level.has_value()) {
+    std::optional<SeverityLevel> ClosestLevel = std::nullopt;
+    for (const auto &[LevelThreshold, _] : PassedLevelsToStreamVariant) {
+      if (LevelThreshold <= Level.value()) {
+        if (ClosestLevel == std::nullopt ||
```

Comment:
```suggestion
        if (ClosestLevel ||
```
The `bool`-ness check will automatically determine if the optional is empty.

Reply:
Yes, but please make sure not to forget the little `!`

Reply:
done